### ISSUE TITLE
Networks/Graphs API Design

### ISSUE THREAD

**Issue by sh4wn** [vispy/vispy#990]
Hi all!

In this issue I would like to discuss the design of the networks/graphs API for Vispy. I know @jvtrudel has got some ideas about that, but maybe other people also have some ideas about this.

Some things to consider:

- NetworkX integration
- Organisation of visuals
- ...

I'll start with this next week, but I think it is a good idea to start a discussion a bit earlier.

**Comment by Eric89GXL** [comment]
I think we should avoid a dependency on NetworkX for viz if possible, at least to start. It's fine to have examples use NetxworkX to e.g. set up a network, but the plotting should only involve a call to a `Vispy` visual / plotting function (that does not depend on NetworkX) in my mind. Is that what you were thinking regarding NetworkX?

**Comment by jvtrudel** [comment]
I do not agree with this strategy. 
NetworkX could be a fast avenue to implement specialized graph transformation. Ultimately, I agree that we should not use NetworkX for "data storage". But it would clearly be an advantage to be able to accept users input  data with the NetworkX formats/objects.

**Comment by Eric89GXL** [comment]
I think it would be okay to eventually have a layer that would accept these types of inputs, but for the first pass I think it's important for orthogonality and low dependency count to have something that works wel for `ndarray`.

**Comment by rossant** [comment]
we can accept networkx objects without a dependency to networkx no?

**Comment by bollu** [comment]
@rossant: I don't really belong to the discussion, but how? AFAIK even ```isinstance(obj, type)``` needs the ```type```. So, how do you check if an object is a networkx object? 

One way maybe by accepting a keyword argument ```GraphVisual(networkxdata=_)``` ?

**Comment by rossant** [comment]
I imagine you can test by checking attributes (duck typing)

**Comment by bollu** [comment]
rossant: hm. so, catch the ```AttributeError``` if you're not able to access a networkx style attribute?

**Comment by Eric89GXL** [comment]
I don't think looking for attributes would be safe / clean enough, but we could have a simple thing like:
```
try:
    from networkx import object_type
except Exception:
    object_type is None
```
Then proceed later to use it appropriately. But in any case, something that worked with general arrays or maybe sparse matrices (scipy support?) would be a prerequisite in my mind.

**Comment by bollu** [comment]
@Eric89GXL this is for rendering _and_ representing graphs, right?

**Comment by Eric89GXL** [comment]
Rendering. I'd prefer to have some API that worked for some simple way of representing graphs, then extract that type of representation from objects (e.g., networkx objects) if possible. If it's not possible to represent the graphs without networkx, then I guess we'll have to add a dependency on it.

**Comment by bollu** [comment]
The discussion reminded me of this:

http://bl.ocks.org/mbostock/4062045

The use a simple ```[node]```, [edge]``` representation for that visualization

**Comment by Eric89GXL** [comment]
That would be an awesome example to replicate here. Similar syntax to what graphviz / pygraphviz uses.

**Comment by bollu** [comment]
Yeah, d3 has crazy awesome graph examples that convinced me to use it in the first place :) A couple more:

- [collapsible graph](http://mbostock.github.io/d3/talk/20111116/force-collapsible.html): Click on the nodes to collapse

- [tree layout](http://bl.ocks.org/mbostock/4063550)

**Comment by rougier** [comment]
There is a simple example using force-directed graph in https://github.com/glumpy/glumpy/blob/master/examples/graph.py

There is no dependency but the parameters need to be tuned a little bit. It is mainly a collection of markers and a collection of segmentS. I've studied networkx and from what I've found, you cannot really use it for for the force-directed animation.

**Comment by almarklein** [comment]
About networkx, we could import it as soon as its functionality is being used. I would not mind to have it as a dependency *for graphs*, since its pure Python.

**Comment by Eric89GXL** [comment]
If I have a graph defined using some simple syntax e.g. for use with `pygraphviz`, I'd rather not have to first go through `networkx` in order to run a `vispy` function. This type of thing, in addition to the standard dependency arguments, is a reason why I'd prefer to have a layer that doesn't require using `networkx`.

**Comment by almarklein** [comment]
So, would vispy have to know about both networkx and pyqtgraphviz?

**Comment by Eric89GXL** [comment]
No, my idea is that `vispy` should know about some simple syntax that is easy to use and get to/from from any other format, assuming such a syntax exists.

**Comment by Eric89GXL** [comment]
This seems more in line with how we currently deal with input data than making our graphing functions support networkx, pygraphviz, etc., at least to at the lowest level / to start.

**Comment by rougier** [comment]
From what I've understood of networkx, you can easily extract a connectivity (n,n) array indicating nodes connections and another (n) array for the node attributes (size, colors, etc.). If vispy provides  just that, it might be more than enough to support networkx without further dependency.

**Comment by sh4wn** [comment]
Yes this is a bit of a dilemma: NetworkX already provides all the necessary datastructures for graphs, and they're also battle tested. But it also provides a lot of functionality we don't need.

## Approaches

1. Depend on NetworkX for the graph plotting (not for the rest of Vispy)

    Advantages: Includes well known and tested datastructures, less work for us, less code maintenance.

    Disadvantages: Quite big dependency, less control on the API

2. Create our own datastructures which provides the minimum we need for plotting 

    Advantages: No dependencies, more control, provides exactly the API we need

    Disadvantages: More work, less tested, extra learning curve for users, performance when need to convert one datastructure to the other?

3. Hybrid: copy the basic datastructures from NetworkX we need for plotting, and add them to the vispy.ext module. Import those if networkx is not installed, otherwise use the installed networkx package.

    Advantages: Well known and tested datastructures, less programming work

    Disadvantages: Possible version conflicts, need to keep track of changes upstream. 

Any suggestions? What option does everybody prefer?

**Comment by rossant** [comment]
as @rougier suggested, can't we just support:

* connectivity matrix as a NumPy array
* dictionary for node attributes
* and optionally, something like

```python
def func_graph(graph=None, matrix=None, attrs=None):
    if hasattr(graph, 'to_matrix'):  # assuming this is a networkX function...
        matrix = graph.to_matrix()
    # ... idem for attrs. we can also try/except a NetworkX import
    # ... then we pursue with matrix and attrs
```

**Comment by sh4wn** [comment]
Yeah actually that's probably a good idea. We'll need some attributes for the edges too, but that's an easy addition. With this we also say "hey we're not a full blown graph library, we only need the connectivity".

**Comment by campagnola** [comment]
Agreed--connectivity, edge attributes, and node attributes can all be represented in numpy arrays. As long as there is a convenient way to generate these arrays using the popular graph libraries, we have no need to support them directly.

It would be nice to support sparse matrix MxM connectivity arrays, but for now it will be easiest to support Nx2 connectivity arrays (also very common in graph computing) because this is exactly the format needed for drawing lines.

**Comment by Eric89GXL** [comment]
+1 for this plan, sounds like we have a reasonable starting point

**Comment by jvtrudel** [comment]
Good, lets work on connectivity, edge attributes, and node attributes. 

[This could be a good reference](https://cs.brown.edu/~rt/gdhandbook/) of graph visualization methods.

**Comment by jvtrudel** [comment]
I did a _very naive_ prototypal implementation of graph _API_ having as ultimate goal a mindMap API. I transformed a networkx graph object in agency matrix (using only the symmetrically relevant part of it as I do not treat directed graph for now). I used an _electrostatic_ method to build the layout [as described in section 2 of this paper](http://arxiv.org/pdf/1201.3011.pdf). Nodes are then rendered very simply using app.canvas. 

I try to separate the graphData (connectivity) from the layout and styles parameters.  I would like to implement different layout algorithms.

I try to design a graph loader that could handle various input formats but networkx already have this feature implemented for many standard format.

You can find [a notebook of this implementation on my vispy's fork](https://github.com/jvtrudel/vispy/blob/master/vispy/mindMapAPI/mindMap.ipynb). I think it would be nice to use collections and sceneGraph.

**Comment by sh4wn** [comment]
Cool links, @jvtrudel! Will be very useful in the near future. 

Today I've been thinking on some use cases and few general requirements of the API. Please let me hear what you think:

Graph API
===========

Initial user story: visualizing biological networks.

Audience:

* **Bio-informaticians**

    - Have good CS knowledge
    - Would like to have flexible API with a customization options
    - Probably embed the functionality of Vispy in an other application
* **Bio-chemists**

    - Less CS knowlegde
    - Want to easily load and visualize graphs from existing sources (pdb.org, KEGG PATHWAY, ...)
    - Probably use example Vispy script without much customization
    - Wants to interact with the visualization (move nodes, extra information on click, ...)

Use Case 1: Protein-Protein interaction networks
-------------------------------------------------------------------

![Example PPI network](http://jcs.biologists.org/content/118/21/4947/F3.large.jpg)

* Huge networks
* Non-directed
* In general node contains most information (metadata)

Use Case 2: Metabolic pathways
---------------------------------------------

![Example metabolic pathway](http://www.genome.jp/kegg/pathway/map/map00120.png)

* Relatively small
* Directed
* Both nodes and edges contain a lot of metadata

Requirements
--------------------

* Should support directed and non directed edges
* Should support multiple edges per pair of nodes
* Should support arbitrary metadata per node
* Should support arbitrary metadata per edge
* Should support multiple types of visuals for nodes
* Should support multiple types of visuals for edges
* Node visual should be configurable based on node metadata
* Edge visual should be configurable based on edge metadata
* Should be able to position nodes programmatically
* Should be able to position edges programmatically
* Should be able to automatically layout nodes/edges
* Should support multiple automatic layouts.
* ...?

Data Flow
-------------

![img_20150629_163453](https://cloud.githubusercontent.com/assets/24530/8410206/fdfd4f7e-1e7c-11e5-93f4-9fb003b97cb8.jpg)

I'm not super sure yet how to draw the graph using a GraphVisual yet. Create MarkerVisual members for each node, and Arrow/LineVisuals for each edge? In that case GraphVisual would be merely a sort of organizing class, and does not do any OpenGL rendering itself.

**Comment by bollu** [comment]
Should the GraphVisual use campagnola/vispy's rewrite of visuals?

In particular, this would be a CompoundVisual if it were using the new
Visuals/scenegraph code.

**Comment by sh4wn** [comment]
Yes it probably should. I should invest some time in investigating the changes in the PR from @campagnola.

**Comment by campagnola** [comment]
> In that case GraphVisual would be merely a sort of organizing class, and does not do any OpenGL rendering itself.

That's generally true of all high-level visuals.
I agree that you should start this as a CompoundVisual.

**Comment by sh4wn** [comment]
Ok, so to start simple:

1. We're going to have a GraphVisual. This is a CompoundVisual containing an ArrowVisual and a MarkersVisual at first. No support for custom colors per node/edge yet. 
2. The GraphVisual is a low level visual, in the sense that this visual will only draw the graph (nodes and edges). Text labels and more fancy stuff will be done by a higher level API.
3. The GraphVisual will contain a Layouter member. Each Layouter subclass implements a graph layouting algorithm, using the Python iterator protocol, which makes it possible to get intermediate results, which allowing us to animate the layouting.

Questions:

1. I want to start with randomly placing the nodes. This should happen inside the GraphVisual class. But can I access the current window size from inside a Visual subclass? Or is there any other way to randomly place the nodes in the current canvas?
2. How shall we organize the code? Possibilities:
    * vispy.visuals.graph package containing the GraphVisual, and another subpackage with the layouters
    * Move the layouters to a subpackage outside vispy.visuals, for example vispy.graphs.layouters.
    * ...?

**Comment by sh4wn** [comment]
Regarding Q1, that's probably doable using Visual.get_transform

**Comment by Eric89GXL** [comment]
1. You shouldn't really need to worry about how big the window is. If you just always construct your graphs e.g. to be unit size (bounded by the -1, +1 box), then it should be pretty easy for people to scale them as they want. Most of the time, users will have them in a scenegraph, where this will be auto-scaled to the right size for the window. This is the way the other visuals work, too -- no auto-scaling based on window size or anything.

2. I think putting everything in `vispy.visuals.graph` is good for now, but make sure `GraphVisual` is imported in `vispy.visuals` itself. That way most people can just get the graph visual from there, but if they need more advanced functionality they can get it from `vispy.visuals.graph`. Eventually it might be possible to do things like `my_graph.layouter = 'circular'` or whatever, and have the string automatically get translated into an instance of the right class, so users won't need to do extra imports. We already do this with cameras in the scenegraph.

Otherwise, this sounds like a good plan to me.

**Comment by sh4wn** [comment]
Ok, I've got a temporary PR here: https://github.com/sh4wn/vispy/pull/23.

Current status:
![screenshot from 2015-07-28 16-34-28](https://cloud.githubusercontent.com/assets/24530/8934206/d6793dd8-3546-11e5-93b3-e27b76c82c3f.png)

This should be a directed graph, but you don't see the arrows because they're hidden behind the nodes. Here's the problem: the node vertices are generated at random, and these (x, y) pairs have values in the range [0, 1]. 

The graph edges should point from node to node of course. But the vertices for the edge line should not lie in the middle of the node, because that's why the arrow heads are now beneath the node. The vertices for the edges should lie before the node, but what would be a good way to calculate these positions (which also depend on the direction vector of the edge)? I can't just substract `marker_size * direction` from the node position, because marker_size is not in the range [0, 1].

**Comment by Eric89GXL** [comment]
You basically need a way to be able to subtract in (logical) pixel units some length from each line. Right now I don't know of a way to do it without modifying the GLSL to take this into account in e.g. the vertex shader. @rougier is that how you would solve this problem?

**Comment by sh4wn** [comment]
Got a simple circular layout working:

![screenshot from 2015-07-31 17-41-41](https://cloud.githubusercontent.com/assets/24530/9011105/72d05cd2-37ab-11e5-9599-95bf4c3ccd89.png)

Shall I create another PR on the VisPy master branch for this GraphVisual?

**Comment by Eric89GXL** [comment]
Yep, but let me merge your current PR first. That way you can `rebase` before opening the PR and be in a clean state.

**Comment by rougier** [comment]
@sh4wn  You'll need to find the exact coordinates of your marker (in pixels) and this depends on the overall transform that is used. @campagnola do you know how to access it from within shader code ?

**Comment by bollu** [comment]
Document corresponds to (logical) pixel coordinates.

It can be accessed as

```
Visual_to_doc = self.transforms.get('visual', 'document')
```

don't forget to convert from document to render coordinates once you're
done using them :)

**Comment by bollu** [comment]
Oh, the transform can then be sent to the shader using the `$shadervar`
method, the same way `$transform ` is sent to the shaders.
